PLANNER-505: adding support for JMS

etirelli opened the pull request:
Adding support for JMS transport. 

@ge0ffrey not sure why the kjar with the planner cloudbalance example is raising compilation failures. It is not related to my changes, as it was already happening. Will investigate tomorrow.

ge0ffrey commented on kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/CreateSolverCommand.java:
```diff
@@ -0,0 +1,88 @@
+        if ( instance == null ) {
+            if ( other.instance != null )
+                return false;
+        } else if ( !instance.equals( other.instance ) )
```
this will hardly ever be equal as SolverInstance doesn't implement equals (so they are only equal if they are the same instance, which is never).

I suggest removing the equals/hashcode code from this Command for now - until it's sure that we need to have such code.

etirelli commented:
@ge0ffrey @mswiderski @sutaakar this is ready for merging. All tests passed locally. Can you please review?

ge0ffrey commented on kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/UpdateSolverStateCommand.java:
```diff
@@ -0,0 +1,108 @@
+
+    @XmlElement
+    @XStreamAlias("solver-instance")
+    private String instance;
```
Why is solver-instance of type String?

etirelli replied:
Because the instance may contain user-provided classes (for instance, the problem to solve). This requires that attribute to be unmarshalled with the kie-container classloader, not the kie-server classloader. Unfortunately, this is the only way we found to do it. It is the same artifice drools/jbpm use.

sutaakar commented on kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/impl/SolverServicesClientImpl.java:
```diff
@@ -100,13 +104,15 @@ public SolverServicesClientImpl(KieServicesConfiguration config, ClassLoader cla
     public ServiceResponse<SolverInstance> updateSolverState(String containerId, String solverId, SolverInstance instance) {
         checkMandatoryParameter( "ContainerID", containerId );
         checkMandatoryParameter( "SolverId", solverId );
+        checkMandatoryParameter( "instance", instance );
```
I think that SolverInstance don't need to be mandatory when terminating solver.

Maybe best solution would be to separate starting and terminating of solvers into separate client methods with own mandatory parameter validations.

etirelli replied:
The update method requires the solver state attribute that is inside the solver instance, so it is mandatory.

I agree that having client methods for specific operations would make it easier to use, but as of today, this is the same across drools/jbpm as well. We can introduce the additional client operations as part of the next version.

sutaakar commented:
+1 for me

mswiderski commented:
changes looks ok to me though we have quite a few test failures related to this changes, or did I missed something?

etirelli commented:
:( I will try to see what's going on. Tests succeeded locally, but I had a bunch of environment issues before.

etirelli commented:
OK, additional fixes in, squashed, rebased and re-submitted. 

@sutaakar @mswiderski @ge0ffrey in case you want to check again, tests should succeed now.

sutaakar commented:
Tests failing because of changed behaviour of REST DELETE request in Kie client.

etirelli commented:
Hi Karel, yes, I fixed it locally, and I am re-running all the tests now.

etirelli commented:
Jenkins, be nice to me! :)

etirelli commented:
YEY, all tests succeeded! :) ready to merge.

mswiderski commented:
+1 to merge

etirelli commented:
merged.
